[GTK] Improve force repaint implementation to reduce the amount of flaky ref tests

carlosgcampos opened the pull request:
#### 57d42a4b3757962b89cc88e7da3ae63ac38eba32
<pre>
[GTK] Improve force repaint implementation to reduce the amount of flaky ref tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=284524">https://bugs.webkit.org/show_bug.cgi?id=284524</a>

Reviewed by Nikolas Zimmermann.

Many ref tests are flaky because we are not properly taking the snapshot
of either the reference or test. This patch reworks the force repaint
implementation of the GTK port, with some parts also shared with WPE.

This patch includes the following changes:

 - For GTK we try to take the snapshot by using the last committed
   buffer, instead of using GTK API to get the web view rendered
   contents. When not possible for whatever reason, we fall back to
   getting web view rendered contents.
 - Send the DispatchAfterEnsuringDrawing message to the web process to
   make sure we get another frame in the UI process with the frame we
   want. This also allows as to remove the timer in GTK port to wait for
   multiple frames, since now we know the next one is the one we want.
 - To make sure we use the new rendered frame the paint/snapshot
   methods return a bool indicating whether buffers were swapped in that
   paint.
 - In the web process force repaint has been reworked to follow what
   other ports do, calling Page::forceRepaintAllFrames() and then doing
   a synchronous layer flush where a composition is forced. In case of
   async force repaint, we wait for the composition triggered by the
   force repaint (using the composition request ID) to notify back the
   UI process.
 - Resize no longer uses force repaint, it either schedules a new layer
   flush if waiting for the renderer, or does a sync layer flush
   otherwise.
 - We don&apos;t really need a sync composition, so ThreadedCompositor::forceRepaint()
   has been removed, since it has always been unimplemented and we don&apos;t
   plan to implement it either.
 - GTK run loop priority for accelerated compositing timers (LayerFlush,
   CompositingThreadUpdate and DisplayRefreshMonitor) has been changed
   to be the same as WebCore timers to make sure a WebCore timer can&apos;t
   prevent a layer flush. This caused a timeout in one test in GTK port.
   This was not a problem in WPE because it already uses the same
   priority for those timers.

* LayoutTests/platform/glib/printing/print-with-media-query-destory-expected.txt: Renamed from LayoutTests/platform/gtk/printing/print-with-media-query-destory-expected.txt.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/wtf/glib/RunLoopSourcePriority.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
(WebCore::CoordinatedBackingStoreProxy::Update::waitUntilPaintingComplete):
(WebCore::CoordinatedBackingStoreProxy::waitUntilPaintingComplete):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
(WebCore::CoordinatedBackingStoreTile::processPendingUpdates):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::waitUntilPaintingComplete):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.cpp:
(WebCore::CoordinatedAcceleratedTileBuffer::serverWait):
(WebCore::CoordinatedAcceleratedTileBuffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(_WebKitWebViewBasePrivate::_WebKitWebViewBasePrivate):
(webkitWebViewBaseNextPresentationUpdateFrame):
(webkitWebViewBaseDispose):
(webkitWebViewBaseSnapshot):
(webkitWebViewBaseDraw):
(webkitWebViewBaseCallAfterNextPresentationUpdate):
(createSkImageFromCairoSurface):
(webkitWebViewBaseSnapshotFromWidget):
(webkitWebViewBaseSnapshotForTesting):
(_WebKitWebViewBasePrivate::nextPresentationUpdateTimerFired): Deleted.
(webkitWebViewBaseNextPresentationUpdateMonitorStart): Deleted.
(webkitWebViewBaseNextPresentationUpdateMonitorStop): Deleted.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBaseInternal.h:
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::dispatchPresentationCallbacksAfterFlushingLayers):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStore.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::rendererBufferMode):
(WebKit::nativeImageFromGdkTexture):
(WebKit::AcceleratedBackingStoreDMABuf::BufferDMABuf::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferSHM::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::snapshot):
(WebKit::AcceleratedBackingStoreDMABuf::paint):
(WebKit::AcceleratedBackingStoreDMABuf::bufferAsNativeImageForTesting const):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::updateRenderingWithForcedRepaint):
(WebKit::DrawingAreaCoordinatedGraphics::updateRenderingWithForcedRepaintAsync):
(WebKit::DrawingAreaCoordinatedGraphics::updateGeometry):
(WebKit::DrawingAreaCoordinatedGraphics::dispatchAfterEnsuringDrawing):
(WebKit::DrawingAreaCoordinatedGraphics::dispatchPendingCallbacksAfterEnsuringDrawing):
(WebKit::DrawingAreaCoordinatedGraphics::display):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::flushLayers):
(WebKit::LayerTreeHost::layerFlushTimerFired):
(WebKit::LayerTreeHost::forceRepaint):
(WebKit::LayerTreeHost::forceRepaintAsync):
(WebKit::LayerTreeHost::ensureDrawing):
(WebKit::LayerTreeHost::sizeDidChange):
(WebKit::LayerTreeHost::didComposite):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::forceRepaint): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
(WTR::viewSnapshot): Deleted.

Canonical link: <a href="https://commits.webkit.org/288306@main">https://commits.webkit.org/288306@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eecd0b053cadbcf56c52bf2ee1c09d2d0d4746f5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2017 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36432 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87485 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9901 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64183 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21934 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1470 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74935 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1371 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32455 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75341 "Found 1 new JSC stress test failure: stress/scoped-arguments-table-should-be-tolerant-for-oom.js.dfg-eager-no-cjit-validate (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72662 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29746 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88841 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81407 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9659 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6917 "Found 1 new test failure: http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72584 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70749 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71801 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15937 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14947 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1018 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12795 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9612 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15133 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103820 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9486 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25196 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12952 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11256 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -87,6 +97,11 @@ OptionSet<DMABufRendererBufferMode> AcceleratedBackingStoreDMABuf::rendererBuffe
         if (forceSHM && strcmp(forceSHM, "0"))
             return;
 
+        // Don't claim to support hardware buffers if we don't have a device to import them.
```
That seems unrelated, to this patch -- while also important to have.

carlosgcampos replied:
Yes, more or less, I had to do that to ensure tests run in swrast mode here when using WTR directly

nikolaszimmermann reviewed (approved):
Looks good to me!

carlosgcampos committed:
[GTK] Improve force repaint implementation to reduce the amount of flaky ref tests
https://bugs.webkit.org/show_bug.cgi?id=284524

Reviewed by Nikolas Zimmermann.

Many ref tests are flaky because we are not properly taking the snapshot
of either the reference or test. This patch reworks the force repaint
implementation of the GTK port, with some parts also shared with WPE.

This patch includes the following changes:

 - For GTK we try to take the snapshot by using the last committed
   buffer, instead of using GTK API to get the web view rendered
   contents. When not possible for whatever reason, we fall back to
   getting web view rendered contents.
 - Send the DispatchAfterEnsuringDrawing message to the web process to
   make sure we get another frame in the UI process with the frame we
   want. This also allows as to remove the timer in GTK port to wait for
   multiple frames, since now we know the next one is the one we want.
 - To make sure we use the new rendered frame the paint/snapshot
   methods return a bool indicating whether buffers were swapped in that
   paint.
 - In the web process force repaint has been reworked to follow what
   other ports do, calling Page::forceRepaintAllFrames() and then doing
   a synchronous layer flush where a composition is forced. In case of
   async force repaint, we wait for the composition triggered by the
   force repaint (using the composition request ID) to notify back the
   UI process.
 - Resize no longer uses force repaint, it either schedules a new layer
   flush if waiting for the renderer, or does a sync layer flush
   otherwise.
 - We don't really need a sync composition, so ThreadedCompositor::forceRepaint()
   has been removed, since it has always been unimplemented and we don't
   plan to implement it either.
 - GTK run loop priority for accelerated compositing timers (LayerFlush,
   CompositingThreadUpdate and DisplayRefreshMonitor) has been changed
   to be the same as WebCore timers to make sure a WebCore timer can't
   prevent a layer flush. This caused a timeout in one test in GTK port.
   This was not a problem in WPE because it already uses the same
   priority for those timers.

* LayoutTests/platform/glib/printing/print-with-media-query-destory-expected.txt: Renamed from LayoutTests/platform/gtk/printing/print-with-media-query-destory-expected.txt.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/wtf/glib/RunLoopSourcePriority.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
(WebCore::CoordinatedBackingStoreProxy::Update::waitUntilPaintingComplete):
(WebCore::CoordinatedBackingStoreProxy::waitUntilPaintingComplete):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
(WebCore::CoordinatedBackingStoreTile::processPendingUpdates):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::waitUntilPaintingComplete):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.cpp:
(WebCore::CoordinatedAcceleratedTileBuffer::serverWait):
(WebCore::CoordinatedAcceleratedTileBuffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(_WebKitWebViewBasePrivate::_WebKitWebViewBasePrivate):
(webkitWebViewBaseNextPresentationUpdateFrame):
(webkitWebViewBaseDispose):
(webkitWebViewBaseSnapshot):
(webkitWebViewBaseDraw):
(webkitWebViewBaseCallAfterNextPresentationUpdate):
(createSkImageFromCairoSurface):
(webkitWebViewBaseSnapshotFromWidget):
(webkitWebViewBaseSnapshotForTesting):
(_WebKitWebViewBasePrivate::nextPresentationUpdateTimerFired): Deleted.
(webkitWebViewBaseNextPresentationUpdateMonitorStart): Deleted.
(webkitWebViewBaseNextPresentationUpdateMonitorStop): Deleted.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBaseInternal.h:
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::dispatchPresentationCallbacksAfterFlushingLayers):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStore.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::rendererBufferMode):
(WebKit::nativeImageFromGdkTexture):
(WebKit::AcceleratedBackingStoreDMABuf::BufferDMABuf::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferGBM::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::BufferSHM::asNativeImageForTesting const):
(WebKit::AcceleratedBackingStoreDMABuf::snapshot):
(WebKit::AcceleratedBackingStoreDMABuf::paint):
(WebKit::AcceleratedBackingStoreDMABuf::bufferAsNativeImageForTesting const):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::callAfterNextPresentationUpdate):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::updateRenderingWithForcedRepaint):
(WebKit::DrawingAreaCoordinatedGraphics::updateRenderingWithForcedRepaintAsync):
(WebKit::DrawingAreaCoordinatedGraphics::updateGeometry):
(WebKit::DrawingAreaCoordinatedGraphics::dispatchAfterEnsuringDrawing):
(WebKit::DrawingAreaCoordinatedGraphics::dispatchPendingCallbacksAfterEnsuringDrawing):
(WebKit::DrawingAreaCoordinatedGraphics::display):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::flushLayers):
(WebKit::LayerTreeHost::layerFlushTimerFired):
(WebKit::LayerTreeHost::forceRepaint):
(WebKit::LayerTreeHost::forceRepaintAsync):
(WebKit::LayerTreeHost::ensureDrawing):
(WebKit::LayerTreeHost::sizeDidChange):
(WebKit::LayerTreeHost::didComposite):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::forceRepaint): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
(WTR::viewSnapshot): Deleted.

Canonical link: https://commits.webkit.org/288306@main
